bpf/lib/nat.h

```diff
@@ -835,15 +934,83 @@ snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
 	case IPPROTO_ICMP:
 		if (ctx_load_bytes(ctx, off, &icmphdr, sizeof(icmphdr)) < 0)
 			return DROP_INVALID;
-		if (icmphdr.type != ICMP_ECHO &&
-		    icmphdr.type != ICMP_ECHOREPLY)
-			return DROP_NAT_UNSUPP_PROTO;
-		if (icmphdr.type == ICMP_ECHO) {
+		switch (icmphdr.type) {
+		case ICMP_ECHO:
 			tuple.dport = 0;
 			tuple.sport = icmphdr.un.echo.id;
-		} else {
+			break;
+		case ICMP_ECHOREPLY:
 			tuple.dport = icmphdr.un.echo.id;
 			tuple.sport = 0;
+			break;
+		case ICMP_DEST_UNREACH:
+			switch (icmphdr.code) {
+			case ICMP_FRAG_NEEDED: {
+				struct iphdr iphdr;
+				__be16 identifier;
+				__u8 type;
+				__u32 icmpoff = off + sizeof(icmphdr);
+
+				/* According to the RFC 5508, any networking
+				 * equipment that is responding with an ICMP Error
+				 * packet should embed the original packet in its
+				 * response.
+				 */
+
+				if (ctx_load_bytes(ctx, icmpoff, &iphdr,
+						   sizeof(iphdr)) < 0)
+					return DROP_INVALID;
+
+				/* From the embedded IP headers we should be able
+				 * to determine corresponding protocol, IP src/dst
+				 * of the packet sent to resolve the NAT session.
+				 */
+				tuple.nexthdr = iphdr.protocol;
```

Comment:
`tuple` was previously initialized using the outer headers in `snat_v4_init_tuple()`, now this same variable is reused to hold the inner tuple. This means that depending on the `icmphdr.type` the `tuple` could represent either the outer or inner tuple. Could that be a problem? If these tuples are somehow different for some reason, it seems like this could trigger mistranslation of the outer header. Should these be tracked in a separate variable to avoid the inner tuple impacting the outer translation?

Reply:
It's a bit tricky, but I'm certainly open to change that part. If I use a different  tuple I will have to add conditions to use one or the other when trying to handle the nat mapping.

I was considering that, what we really want to rev-nat is the embedded headers, the outer headers are not really needed as we don't want to do anything with them.